File: src/01_data_structures/02_linked_list/implementations/ts/linked_list.ts
type ListNode<T> = {
  value: T;
  next: ListNode<T> | null;
};

interface ILinkedList<T> {
  prepend(value: any): this;
  append(value: any): this;
  remove(matcher: (arg: T) => boolean): this;
  find(matcher: (arg: T) => boolean): boolean;
}

export class LinkedList<T> implements ILinkedList<T> {
  head: ListNode<T> = null;

  prepend(value: any): this {
    // Create new node with value and head as next and prepend it
    this.head = {
      value,
      next: this.head,
    };

    return this;
  }

  append(value: any): this {
    // If list is empty, create and add node
    if (this.head === null) {
      this.head = {
        value,
        next: null,
      };
      return this;
    }

    // Iterate list
    let current = this.head;
    while (current.next !== null) {
      current = current.next;
    }

    // At the end, append new node with value
    current.next = { value, next: null };

    return this;
  }

  remove(matcher: (arg: T) => boolean): this {
    // If list empty, return
    if (this.head === null) {
      return this;
    }

    // If first item, remove and return
    if (matcher(this.head.value)) {
      this.head = this.head.next;

      return this;
    }

    // Iterate list
    let current = this.head;
    while (current.next !== null) {
      if (matcher(current.next.value)) {
        // Remove the node by skipping it
        current.next = current.next.next;

        return this;
      }

      current = current.next;
    }

    return this;
  }

  find(matcher: (arg: T) => boolean): boolean {
    let current = this.head;

    while (current !== null) {
      // If found
      if (matcher(current.value)) {
        return true;
      }
      current = current.next;
    }

    // Not found
    return false;
  }
}

File: src/01_data_structures/02_linked_list/implementations/go/linked_list.go
package linkedList

type Node[T comparable] struct {
	Value T
	Next  *Node[T]
}

type ILinkedList[T comparable] interface {
	Prepend(value T) *LinkedList[T]
	Append(value T) *LinkedList[T]
	Remove(matcher func(T) bool) *LinkedList[T]
	Find(matcher func(T) bool) bool
}

type LinkedList[T comparable] struct {
	Head *Node[T]
}

func (list *LinkedList[T]) Prepend(value T) *LinkedList[T] {
	// Create new node with value and head as next and prepend it
	list.Head = &Node[T]{
		Value: value,
		Next:  list.Head,
	}

	return list
}

func (list *LinkedList[T]) Append(value T) *LinkedList[T] {
	// If list is empty, create and add node
	if list.Head == nil {
		list.Head = &Node[T]{Value: value, Next: nil}

		return list
	}

	// Iterate list
	current := list.Head
	for current.Next != nil {
		current = current.Next
	}

	// At the end, append new node with value
	current.Next = &Node[T]{
		Value: value,
		Next:  nil,
	}

	return list
}

func (list *LinkedList[T]) Remove(matcher func(T) bool) *LinkedList[T] {
	// If list empty, return
	if list.Head == nil {
		return list
	}

	// If first item, remove and return
	if matcher(list.Head.Value) {
		list.Head = list.Head.Next

		return list
	}

	// Iterate list
	current := list.Head
	for current.Next != nil && !matcher(current.Next.Value) {
		current = current.Next
	}

	// If last item, return
	if current.Next == nil {
		return list
	}

	// Item found, link current to next next item and return
	current.Next = current.Next.Next

	return list
}

func (list *LinkedList[T]) Find(matcher func(T) bool) bool {
	// Iterate list
	current := list.Head
	for current != nil {
		// If value found, return true
		if matcher(current.Value) {
			return true
		}

		current = current.Next
	}

	// If not found, return false
	return false
}
